### DISCUSSION TITLE
Easy way to show signature for many functions

### DISCUSSION THREAD

**Discussion by sun-rs** [PyO3/pyo3#3021] [Questions]
```rust
m.add_function(wrap_pyfunction!(ve::func1, m)?)?;
m.add_function(wrap_pyfunction!(ve::func2, m)?)?;
m.add_function(wrap_pyfunction!(ve::func3, m)?)?;
```
```python
def func1(nd: Date) -> int:
    """Convert [`NaiveDate`] to [`i32`]"""
    return api.nfunc1(nd)
```

I have wrapped many function(more than 100) from rust to python extension, and I found it is hard to remember these function signatures(arguments number and type/output type/etc). So I wrapped these builtin functions to a new function in python one by one(with arguments detail and doc). Is these some easy way to get some doc or signature info?

**Comment by davidhewitt** [comment]
You should take a look at http://pyo3.rs/latest/function/signature - the `__text_signature__` on all the Python functions can give you all information except for type annotations.

For type annotations, the latest discussion is in #2454

**Reply by sun-rs** [reply]
Thanks for reply. With `#[pyo3(text_signature = some _info)]` I could get the signature info and with `/// docstring` in rust I could get the same docstring in python.

But what I really confused is that in Vscode, when move mouse to the function wrapped by pyo3, it only shows `function_name: Any`. But move to a python function(not builtin) would show more info like signature and docstring. This cause people don't know the docstring and signature when edit a .py file and also don't know exactly how many `#[pyfunction]` and `#[pyclass]` are wrapped in the extension. 

I know this is not a pyo3 problem, but is there any solution to make this remember process easier?

**Reply by kngwyu** [reply]
Maybe we need to implement #510 to enable that?
